src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorDiagnosticsEndpoint.cs

```diff
@@ -183,12 +183,27 @@ private static bool ShouldFilterHtmlDiagnosticBasedOnErrorCode(OmniSharpVSDiagno
 
             return diagnostic.Code.Value.String switch
             {
+                CSSErrorCodes.MissingOpeningBrace => IsCSharpInStyleBlock(diagnostic, sourceText, syntaxTree),
+                CSSErrorCodes.MissingSelectorAfterCombinator => IsCSharpInStyleBlock(diagnostic, sourceText, syntaxTree),
+                CSSErrorCodes.MissingSelectorBeforeCombinatorCode => IsCSharpInStyleBlock(diagnostic, sourceText, syntaxTree),
                 HtmlErrorCodes.UnexpectedEndTagErrorCode => IsHtmlWithBangAndMatchingTags(diagnostic, sourceText, syntaxTree),
                 HtmlErrorCodes.InvalidNestingErrorCode => IsInvalidNestingWarningWithinComponent(diagnostic, sourceText, syntaxTree),
                 HtmlErrorCodes.MissingEndTagErrorCode => FileKinds.IsComponent(syntaxTree.Options.FileKind), // Redundant with RZ9980 in Components
                 _ => false,
             };
 
+            static bool IsCSharpInStyleBlock(OmniSharpVSDiagnostic d, SourceText sourceText, RazorSyntaxTree syntaxTree)
+            {
+                // C# in a style block causes diagnostics because the HTML background document replaces C# with "~"
+                var owner = syntaxTree.GetOwner(sourceText, d.Range.Start);
+
+                var element = owner.FirstAncestorOrSelf<MarkupElementSyntax>(
+                    n => n.StartTag.Name.Content.Equals("style", StringComparison.Ordinal));
+                var cSharp = owner.FirstAncestorOrSelf<CSharpCodeBlockSyntax>();
+
+                return element.Body.Any(c => c is CSharpCodeBlockSyntax) || cSharp is not null;
```

Comment:
Why do you need to do this piece: `element.Body.Any(c => c is CSharpCodeBlockSyntax)`?

Reply:
Sometimes the diagnostic gets reported on `MarkupTextLiteral` objects before or after things instead of on the C# block itself. So instead we check the whole style block to see if it's got a CSharpBlock and ignore things in that case.